src/transports/ipc.rs

```diff
@@ -0,0 +1,376 @@
+  Writing {
+    buffer: Vec<u8>,
+    current_pos: usize
+  },
+}
+
+/// Writing part of the IPC transport
+/// Awaits new requests using `mpsc::Receiver` and writes them to the socket.
+struct WriteStream {
+  write: WriteHalf<UnixStream>,
+  incoming: mpsc::Receiver<String>,
+  state: WriteState
+}
+
+impl Future for WriteStream {
+  type Item = ();
+  type Error = ();
+
+  fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+    let new_state = match self.state {
+      WriteState::WaitingForRequest => {
+        // Ask for more to write
+        let to_send = try_ready!(self.incoming.poll());
+        if let Some(to_send) = to_send {
+          trace!("Got new message to write: {:?}", to_send);
+          WriteState::Writing {
+            buffer: to_send.into_bytes(),
+            current_pos: 0,
+          }
+        } else {
+          return Ok(futures::Async::NotReady);
+        }
+      },
+      WriteState::Writing { ref buffer, ref mut current_pos } => {
+        try_ready!(Ok(self.write.poll_write()));
+
+        // Write everything in the buffer
+        while *current_pos < buffer.len() {
+          let n = try_nb!(self.write.write(&buffer[*current_pos..]));
+          *current_pos += n;
+          if n == 0 {
+            warn!("Zero write.");
+            return Err(()); // zero write?
+          }
+        }
+
+        WriteState::WaitingForRequest
+      },
+    };
+
+    self.state = new_state;
+    self.poll()
+  }
+}
+/// Reading part of the IPC transport.
+/// Reads data on the socket and tries to dispatch it to awaiting requests.
+struct ReadStream {
+  read: ReadHalf<UnixStream>,
+  incoming: mpsc::Receiver<(usize, Pending)>,
+  pending: BTreeMap<usize, Pending>,
+  buffer: Vec<u8>,
+  current_pos: usize,
+}
+
+impl Future for ReadStream {
+  type Item = ();
+  type Error = ();
+
+  fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+    // Check new incoming response channels
+    let response = self.incoming.poll();
+    if let Ok(futures::Async::Ready(Some((id, response)))) = response {
+      self.pending.insert(id, response);
+    }
+
+    const DEFAULT_BUF_SIZE: usize = 4096;
+    let mut new_write_size = 128;
+    loop {
+      // Read pending responses
+      try_ready!(Ok(self.read.poll_read()));
+
+      if self.current_pos == self.buffer.len() {
+        if new_write_size < DEFAULT_BUF_SIZE {
+          new_write_size *= 2;
+        }
+        self.buffer.resize(self.current_pos + new_write_size, 0);
+      }
+
+      let read = try_nb!(self.read.read(&mut self.buffer[self.current_pos..]));
+      if read == 0 {
+        return Ok(futures::Async::NotReady);
+      }
+
+      let min = self.current_pos;
+      self.current_pos += read;
+      if let Some((output, len)) = Self::extract_response(&self.buffer[0..self.current_pos], min) {
+        // copy rest of buffer to the beginning
+        for i in len..self.current_pos {
+          self.buffer.swap(i, i - len);
+        }
```

Comment:
`ptr::copy` (_not_ non-overlapping) could definitely make this into as few copies as possible. But odds are it will all fit into the L1 cache anyway and won't be a bottleneck.